File: SPOJ/BOTTOM.cpp
#include <bits/stdc++.h>
using namespace std;
 
const int N=5000;
vector<int> adj[N];
int cntSCC,tim,n;
stack<int> st;
int dfs_num[N],dfs_low[N],comp[N];
bool inSCC[N];
 
void SCC(int u){
	dfs_num[u]=dfs_low[u]=++tim;
	st.push(u);
	for(int i=0;i<adj[u].size();i++){
		int v=adj[u][i];
		if(dfs_num[v]==0)
			SCC(v);
		if(!inSCC[v])
			dfs_low[u]=min(dfs_low[u], dfs_low[v]);
	}
 
	if(dfs_low[u]==dfs_num[u]){
		while(true){
			int x=st.top();
			st.pop();
			inSCC[x]=true;
			comp[x]=cntSCC;
			if(x==u)
				break;
		}
		cntSCC++;
	}
}
 
void SCC(){
	for(int i=0;i<n;i++)
		if(dfs_num[i]==0)
			SCC(i);
}
 
int main(){
	while(1){
		scanf("%d",&n);
		if(n==0)
			break;
		int m;
		scanf("%d",&m);
		int x[m];
		int y[m];
		for(int i=0;i<n;i++)
			adj[i].clear();
		for(int i=0;i<m;i++){
			scanf("%d",&x[i]);
			scanf("%d",&y[i]);
			x[i]--;
			y[i]--;
			adj[x[i]].push_back(y[i]);
		}
		memset(dfs_low,0,sizeof dfs_low);
		memset(dfs_num,0,sizeof dfs_num);
		memset(comp,0,sizeof comp);
		memset(inSCC,false,sizeof inSCC);
		SCC();
		int compDeg[cntSCC];
		memset(compDeg,0,sizeof compDeg);
		for(int i=0;i<m;i++){
			int c1=comp[x[i]];
			int c2=comp[y[i]];
			if(c1!=c2)
				compDeg[c1]++;
		}
		vector<int> ans;
		for(int i=0;i<n;i++)
			if(compDeg[comp[i]]==0)
				ans.push_back(i+1);
		bool f=true;
		for(int i=0;i<ans.size();i++){
			int x=ans[i];
			if(!f)
				printf(" ");
			f=false;
			printf("%d",x);
		}
		printf("\n");
	}
}

File: JAVA/SPOJ/BOTTOM.java

import java.io.*;
import java.util.*;
public class BOTTOM {
	public static void main(String[] args)throws Throwable {
		MyScanner sc=new MyScanner();
		PrintWriter pw=new PrintWriter(System.out);
		
		while(true){
			n=sc.nextInt();
			if(n==0)
				break;
			adj=new ArrayList [n];
			for(int i=0;i<n;i++)
				adj[i]=new ArrayList<Integer>();
			int m=sc.nextInt();
			int [] x=new int [m];
			int [] y=new int [m];
			for(int i=0;i<m;i++){
				x[i]=sc.nextInt()-1;
				y[i]=sc.nextInt()-1;
				adj[x[i]].add(y[i]);
			}
			SCC();
			compDeg=new int [cntSCC];
			for(int i=0;i<m;i++){
				int c1=comp[x[i]];
				int c2=comp[y[i]];
				if(c1!=c2)
					compDeg[c1]++;
			}
			ArrayList<Integer> ans=new ArrayList<Integer>();
			for(int i=0;i<n;i++)
				if(compDeg[comp[i]]==0)
					ans.add(i+1);
			boolean f=true;
			for(int i : ans){
				if(!f)
					pw.print(" ");
				f=false;
				pw.print(i);
			}
			pw.println();
		}
		
		pw.flush();
		pw.close();
	}
	static int n,cntSCC;
	static ArrayList<Integer> [] adj;
	static Stack<Integer> stack;
	static int time;
	static int [] dfs_num,dfs_low,comp,compDeg;
	static boolean [] inSCC;
	
	public static void SCC(){
		stack=new Stack<Integer>();
		dfs_num=new int [n];
		dfs_low=new int [n];
		comp=new int [n];
		inSCC=new boolean [n];
		for(int i=0;i<n;i++)
			if(dfs_num[i]==0)
				SCC(i);
	}
	
	public static void SCC(int u){
		dfs_num[u]=dfs_low[u]=++time;
		stack.push(u);
		
		for(int v : adj[u]){
			if(dfs_num[v]==0)
				SCC(v);
			if(!inSCC[v])
				dfs_low[u]=Math.min(dfs_low[u], dfs_low[v]);
		}
		
		if(dfs_low[u]==dfs_num[u]){
			while(true){
				int x=stack.pop();
				inSCC[x]=true;
				comp[x]=cntSCC;
				if(x==u)
					break;
			}
			cntSCC++;
		}
	}
	
	static class MyScanner {
		BufferedReader br;
		StringTokenizer st;
		public MyScanner() {br = new BufferedReader(new InputStreamReader(System.in));}
		String next() {while (st == null || !st.hasMoreElements()) {
			try {st = new StringTokenizer(br.readLine());}
			catch (IOException e) {e.printStackTrace();}}
		return st.nextToken();}
		int nextInt() {return Integer.parseInt(next());}
		long nextLong() {return Long.parseLong(next());}
		double nextDouble() {return Double.parseDouble(next());}
		String nextLine(){String str = "";
		try {str = br.readLine();}
		catch (IOException e) {e.printStackTrace();}
		return str;}
	}
}